docs/build/smart-contracts/example-contracts/bls-signature.mdx

```diff
@@ -0,0 +1,531 @@
+    },
+    KeyPair {
+        sk: hex!("0ce3cd1dcaecf002715228aeb0645c6a7fd9990ace3d79515c547dac120bb9f7"),
+        pk: hex!("19f7e9dcd4ce2bef92180b60d0c7c7b48b1924a36f9fbb93e9ecb8acb3219e26033b83facd4dc6d2e3f9fa0fffafeca8168bd4824e31dc9dfd977fbf037210508bc807c1a6d20f98a044911f6b689328f3f25dd35a6c05e8c6ac3ac6ef0def91"),
+    },
+    KeyPair {
+        sk: hex!("6b4b27ba3ffc953eff3b974142cdac75f98c8c4ab26f93d5adfd49da5d462c3f"),
+        pk: hex!("15f55ec5572026d6c3c7c62b3ce3c5d7539045e9f492f2b1b0860c0af5f5f6b34531dfe4626a92d5c23ac6ad44330cf40e63a8a7234edbb41539c5484eff2cd23b2f0d502a7fd74501b1a05ffee29b24e79cb1ee9fb9b804d84f486283101ee0"),
+    },
+];
+```
+
+We aggregate the signer public keys, by first converting the bytes into `G1Affine`, then add them all up.
+
+```rust
+fn aggregate_pk_bytes(env: &Env) -> BytesN<96> {
+    let bls = env.crypto().bls12_381();
+    let mut agg_pk = G1Affine::from_bytes(BytesN::from_array(env, &KEY_PAIRS[0].pk));
+    for i in 1..KEY_PAIRS.len() {
+        let pk = G1Affine::from_bytes(BytesN::from_array(env, &KEY_PAIRS[i].pk));
+        agg_pk = bls.g1_add(&agg_pk, &pk);
+    }
+    agg_pk.to_bytes()
+}
+```
+
+To produce the signature, the message will first be hashed into G2 via `bls.hash_to_g2`. Here we use our own defined DST.
+
+To aggregate the signatures, we first produce individual signatures by having each signer sign the message. This means multiplying the secret key by the message's G2 point. Then we add them all up. Here we use `g2_msm`, by an array of message (`Vec<G2Affine>`) points and an the array of secret keys (`Vec<Fr>`) and it computes their inner-product which is the aggregate signature we want.
+
+```rust
+const DST: &str = "BLSSIG-V01-CS01-with-BLS12381G2_XMD:SHA-256_SSWU_RO_";
+
+fn sign_and_aggregate(env: &Env, msg: &Bytes) -> BytesN<192> {
+    let bls = env.crypto().bls12_381();
+    let mut vec_sk: Vec<Fr> = vec![env];
+    for kp in KEY_PAIRS {
+        vec_sk.push_back(Fr::from_bytes(BytesN::from_array(env, &kp.sk)));
+    }
+    let dst = Bytes::from_slice(env, DST.as_bytes());
+    let msg_g2 = bls.hash_to_g2(&msg, &dst);
+    let vec_msg: Vec<G2Affine> = vec![
+        env,
+        msg_g2.clone(),
+        msg_g2.clone(),
+        msg_g2.clone(),
+        msg_g2.clone(),
+        msg_g2.clone(),
+        msg_g2.clone(),
+        msg_g2.clone(),
+        msg_g2.clone(),
+        msg_g2.clone(),
+        msg_g2.clone(),
+    ];
+    bls.g2_msm(vec_msg, vec_sk).to_bytes()
+}
+```
+
+Running this test will produce the following budget output, the total cpu consumption for signature verification is around 31M. And you can add as many additional public keys as you like. In general `pairing_check` is a function with linear cost, so the more unique messages that needs to be signed, the higher cost. Here because all signers sign the same content (hash of the call stack of this contract), we can do this in constant time.
+
+```
+---- test::test stdout ----
```

Comment:
Would it make sense to omit all the cost type rows here and just leave the non-zero BLS rows?